### PR TITLE
Speed up `bowtie filter-implementations` with no arguments

### PR DESCRIPTION
**Pull request by adwait-godbole**
https://github.com/bowtie-json-schema/bowtie/assets/68469605/03d0412f-a1a1-45d5-8990-c804c4e2e352

@Julian I am not sure why the test I wrote is failing though :(. Its some `process.returncode` of 78. Does it have something to do with the `ls implementations/` that's happening in the background ?. The command does work on the CLI without starting any Docker containers.

Also shouldn't we output directly without starting the containers even if it was just `bowtie filter-implementations -i python-jsonschema` with no arguments for e.g. ?

----
📚 Documentation preview 📚: https://bowtie-json-schema--1020.org.readthedocs.build/en/1020/

### CODE REVIEW

File: bowtie/_cli.py [hunk 1]
```diff
@@ -986,6 +992,11 @@ async def filter_implementations(
     """
     Output implementations matching a given criteria.
     """
+    if languages == KNOWN_LANGUAGES:
```
Comment (Julian) [inline bowtie/_cli.py:995]: I haven't tried to run this but it looks suspicious.

Doesn't it ignore the other filter? I.e. if you provide a dialect filter doesn't this now output the wrong answer as it will output implementations regardless of whether they support the given dialects.

Comment (adwait-godbole) [reply bowtie/_cli.py:995]: Yeah I did think of this and was going to do it as `if not dialects and languages == KNOWN_LANGUAGES` but the `not dialects` check here was redundant since I've made it sure in the `callback` that this won't happen because the callback straight away returns KNOWN_LANGUAGES and doesn't really perform the below logic:

```python
  LANGUAGE_ALIASES.get(each, each)  # type: ignore[reportUnknownArgumentType]
  for each in value or KNOWN_LANGUAGES
```

But still to make the code more readable I can add the dialect check as well i.e. `if not dialects and languages == KNOWN_LANGUAGES`

Comment (adwait-godbole) [reply bowtie/_cli.py:995]: > Doesn't it ignore the other filter? I.e. if you provide a dialect filter doesn't this now output the wrong answer as it will output implementations regardless of whether they support the given dialects.

Also the tests would've failed in that case

Comment (adwait-godbole) [reply bowtie/_cli.py:995]:
```python
        KNOWN_LANGUAGES
        if not value
        and not ctx.params["dialects"]
```

in here in the `callback` the `not dialects` check happens.

Comment (Julian) [reply bowtie/_cli.py:995]: Got it -- I think it'd be slightly clearer for this check to happen in one place? Do you see an easy way to do that bit here too?

It'd make sense to do it in the callback if we didn't have to do any `if` here but you're doing one anyways to check against `KNOWN_LANGUAGES`.

**Comment by adwait-godbole** [comment]
> @Julian I am not sure why the test here is failing though :(. Its some process.returncode of 78. Does it have something to do with the ls implementations/ that's happening in the background ?. The command does work on the CLI without starting any Docker containers.

> Also shouldn't we output directly without starting the containers even if it was just bowtie filter-implementations -i python-jsonschema with no arguments for e.g. ?

Do you have any comments on the above ?

**Comment by Julian** [comment]
> Also shouldn't we output directly without starting the containers even if it was just bowtie filter-implementations -i python-jsonschema with no arguments for e.g. ?

Yeah that sounds smart -- basically instead of checking for "is it all the implementations" just check whether any other filter is set I guess.

> Its some process.returncode of 78. 

That means the command is failing, 78 is `EX_CONFIG` so it's failing for that reason. If you see no other output it probably points to us being missing some error handling? Does it fail locally?

**Comment by adwait-godbole** [comment]
> Yeah that sounds smart -- basically instead of checking for "is it all the implementations" just check whether any other filter is set I guess.

Cool! I just pushed these changes. Please have a look!

> That means the command is failing, 78 is EX_CONFIG so it's failing for that reason. If you see no other output it probably points to us being missing some error handling? Does it fail locally?

yes its failing locally as well.

**Comment by adwait-godbole** [comment]
@Julian this time while running the returncode 78 is gone now. But when running `bowtie filter-implementations` in the tests environment I think its unable to get the `stdout`.

**Comment by Julian** [comment]
> But when running bowtie filter-implementations in the tests environment I think its unable to get the stdout for some odd reason. 

It's getting stdout fine, stdout is empty -- a debugger is your friend here to remind yourself why, but the answer is that recall the default is different when `sys.stdin` is a tty or not. Simply put when a "human" probably typed the command we default to `Implementation.known`, but when sys.stdin is not a tty that means someone is probably piping something into the command, and there remember we default to reading from stdin.

In other words what you're testing right now is the equivalent of running `echo | bowtie filter-implementations` where we print nothing. You could I guess add some implementations on stdin and test that behavior (that we filter those on stdin without starting up any containers). If you want to test the TTY default, you could try using `pty.spawn` (which is how you create a fake TTY) or use `subprocess.run` directly to connect to the stdin of the test runner, though with the latter probably we will want to skip the test if stdin of that itself is not a TTY.

File: bowtie/_cli.py [hunk 2]
```diff
@@ -970,22 +970,33 @@ def validate(
     "-l",
     "languages",
     type=click.Choice(sorted(KNOWN_LANGUAGES), case_sensitive=False),
-    callback=lambda _, __, value: frozenset(  # type: ignore[reportUnknownLambdaType]
-        LANGUAGE_ALIASES.get(each, each)  # type: ignore[reportUnknownArgumentType]
-        for each in value or KNOWN_LANGUAGES  # type: ignore[reportUnknownArgumentType]
+    callback=lambda _, __, value: (  # type: ignore[reportUnknownLambdaType]
+        KNOWN_LANGUAGES
```
[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: Ig its not really equivalent since we no longer put `KNOWN_LANGUAGES` through the for loop so they won’t get aliased and only then we would be able to compare `languages == KNOWN_LANGUAGES`

[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: If they get aliased then the equality won’t work.

**Comment by adwait-godbole** [comment]
> that we filter those on stdin without starting up any containers

Is there a way of knowing indeed no containers were started ?

File: bowtie/_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: Ig its not really equivalent since we no longer put KNOWN_LANGUAGES through the for loop so they won’t get aliased and only then we would be able to compare languages == KNOWN_LANGUAGES

[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: If they get aliased then the equality won’t work.

**Comment by Julian** [comment]
Sorry I didn't mean to emphasize that part -- some day we'll handle that by having the test suite run without internet connection, but right now the tests *do* actually hit the network in order to download e.g. the `alpine` base image (but only the first time they're run).

For now it's fine that we simply "know" it doesn't hit the network otherwise the test suite will be noticeably slow.

File: bowtie/_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: Did this make sense @Julian ?

**Comment by adwait-godbole** [comment]
@Julian I just pushed the new test and its working now. Although this is my first time working with `pty.spawn`, so let me know your thoughts 😅

**Comment by adwait-godbole** [comment]
@Julian also let me know your thoughts on this one :)

File: bowtie/_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Julian) [reply bowtie/_cli.py:974]: I'll look and see what you mean, but I'd expect there to be two "things" -- "all the known languages" and then "a mapping of aliases to canonical names", and checking whether we are not filtering languages should basically mean checking whether the languages we're using are "all the known languages" *after* de-aliasing, because at that point all of them should be being referred to via their canonical name.

[the remark this answers is not in the document]
Comment (Julian) [reply bowtie/_cli.py:974]: Oh I see, at the minute `KNOWN_LANGUAGES` isn't the "all known languages with their canonical names", because it's also being used for the `Choice`.

File: tests/test_integration.py [hunk 3]
```diff
@@ -1255,6 +1257,37 @@ async def test_info_unsuccessful_start(succeed_immediately):
     assert "failed to start" in stderr.lower(), stderr
 
 
+@pytest.mark.asyncio
+async def test_filter_implementations_no_arguments():
+    output_buffer = BytesIO()
+    stdout, stderr = [], ""
+
+    def read(fd):
```
Comment (Julian) [inline tests/test_integration.py:1265]: This looks too low-level, I'd expect something to exist which does it already, but I'll have to look what that is.

Comment (adwait-godbole) [reply tests/test_integration.py:1265]: I'll have a look

File: bowtie/_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: So are you fine with this way ?

File: tests/test_integration.py (hunk 3 above)
Comment (adwait-godbole) [reply tests/test_integration.py:1265]: I remember now why I had tried doing it this way. The return type of `pty.spawn` ig if I am not wrong is the `pid` and not the `fd (file_descriptor)`. So `pty.spawn` requires a callback which gets the passed `file_descriptor` and when I had tried to not use `select.select` (polling) earlier, for some odd reason the callback was only reading partial data and always there was some missing data. That's why I tried introducing polling using `select.select` to wait for a split second until it gets the data and only then write to the buffer. This way there was no case of any missed data. I had also tried it using `pty.fork` and if I am not wrong and recalling correctly in that case there was no need of polling (I will try it again).

I might be wrong here / there might be some other way but whenever I tried with `pty.spawn` and no polling it was resulting in missing data :(

Comment (Julian) [reply tests/test_integration.py:1265]: Try using the [pexpect module](https://pexpect.readthedocs.io/en/stable/overview.html) with `.expect(pexpect.EOF)` I think (add that to our test requirements).

File: bowtie/_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adwait-godbole) [reply bowtie/_cli.py:974]: I tried this but it doesn't really work since returning `None if not value` is not really the ideal way for e.g. if someone executes
`bowtie filter-implementations -d 2020` 

then in this case `if not value` is **True** so we should be defaulting to returning all the `KNOWN_LANGUAGES` (for comparison purposes (languages == KNOWN_LANGUAGES) and not even aliasing it) and not return `None` here.

[the remark this answers is not in the document]
Comment (Julian) [reply bowtie/_cli.py:974]: Got it, ok, then let's leave it for now.

File: tests/test_integration.py (hunk 3 above)
Comment (adwait-godbole) [reply tests/test_integration.py:1265]: Okay! I will give it a try 👍

**Comment by adwait-godbole** [comment]
@Julian I added `pexpect`. Working smoothly now 👍. No more polling required :)

**Comment by Julian** [comment]
Great thanks!
